### PR TITLE
Add Watch method to health check service

### PR DESCRIPTION
**Pull request by ericgribkoff**
This implements the health check service `Watch` method as specified in https://github.com/grpc/proposal/blob/master/A17-client-side-health-checking.md

### CODE REVIEW

File: src/python/grpcio_tests/tests/health_check/_health_servicer_test.py [hunk 1]
```diff
@@ -46,37 +62,155 @@ def tearDown(self):
...
+            target=_consume_responses, args=(rendezvous, response_queue))
+        thread.start()
+
+        response = response_queue.get(timeout=_QUEUE_TIMEOUT_S)
+        self.assertEqual(health_pb2.HealthCheckResponse.SERVICE_UNKNOWN,
+                         response.status)
+
+        self._servicer.set('some-other-service',
+                           health_pb2.HealthCheckResponse.SERVING)
+        with self.assertRaises(queue.Empty) as context:
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/health_check/_health_servicer_test.py]: `context` seems not used

File: src/python/grpcio_tests/tests/health_check/_health_servicer_test.py [hunk 2]
```diff
@@ -22,21 +23,36 @@
 
 from tests.unit import test_common
 
+from six.moves import queue
+
+_QUEUE_TIMEOUT_S = 5
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/health_check/_health_servicer_test.py]: nit: How about using the constants from `tests.unit.framework.common.test_constants`

File: src/python/grpcio_health_checking/grpc_health/v1/health.py [hunk 3]
```diff
@@ -23,30 +23,92 @@
...
+        self._condition = threading.Condition()
+        self._responses = list()
+        self._open = True
+
+    def __iter__(self):
+        return self
+
+    def _next(self):
+        with self._condition:
+            while not self._responses and self._open:
```
Comment (lidizheng) [inline src/python/grpcio_health_checking/grpc_health/v1/health.py]: Optional: In current implementation, only newly added status or 'close' being called will awake this thread. It is safe to use `if` here.

Comment (ericgribkoff) [reply src/python/grpcio_health_checking/grpc_health/v1/health.py]: Acknowledged

File: src/python/grpcio_health_checking/grpc_health/v1/health.py [hunk 4]
```diff
@@ -23,30 +23,92 @@
...
+            if service not in self._watchers:
+                self._watchers[service] = set()
+            self._watchers[service].add(watcher)
+            context.add_callback(self._on_close_callback(watcher, service))
+        return watcher
+
     def set(self, service, status):
         """Sets the status of a service.
 
     Args:
```
[the remark this answers is not in the document]
Comment (ericgribkoff) [reply src/python/grpcio_health_checking/grpc_health/v1/health.py]: Fixed

File: src/python/grpcio_tests/tests/health_check/_health_servicer_test.py (hunk 2 above)
Comment (ericgribkoff) [reply src/python/grpcio_tests/tests/health_check/_health_servicer_test.py]: Done

File: src/python/grpcio_tests/tests/health_check/_health_servicer_test.py [hunk 5]
```diff
@@ -46,37 +62,155 @@ def tearDown(self):
...
-            service='grpc.test.TestServiceNotServing')
+    def test_check_not_serving_service(self):
+        request = health_pb2.HealthCheckRequest(service=_NOT_SERVING_SERVICE)
         resp = self._stub.Check(request)
         self.assertEqual(health_pb2.HealthCheckResponse.NOT_SERVING,
                          resp.status)
 
-    def test_not_found_service(self):
+    def test_check_not_found_service(self):
         request = health_pb2.HealthCheckRequest(service='not-found')
```
[the remark this answers is not in the document]
Comment (ericgribkoff) [reply src/python/grpcio_tests/tests/health_check/_health_servicer_test.py]: The string is only used here, so I think it's ok as-is

File: src/python/grpcio_tests/tests/health_check/_health_servicer_test.py (hunk 1 above)
Comment (ericgribkoff) [reply src/python/grpcio_tests/tests/health_check/_health_servicer_test.py]: Removed

**Pull request merged**